node/cancellation-v2.md

```diff
@@ -0,0 +1,139 @@
+          await timer;
+        });
+        cancel(promise);
+        await promise;
+      }
+      ```
+
+## Changes to the current design
+
+### Naming and structure
+
+To solve issue (1), in this proposal the term "scope" goes away and blends in with the concept of Workflow `Context`.
+
+```
+cancellationScope -> Context.cancellable
+shield -> Context.nonCancellable
+cancel -> ContextualPromise.cancel (not related to (1), more on this below)
+```
+
+### New concepts
+
+With these new concepts we can address issue (2).
+
+#### `ContextualPromise` (or `CancellablePromise`)
+
+- Implements the `PromiseLike` interface (awaitable via `then()`), exposes the `cancel` method
+- Context aware: linked to parent context and propagates cancellation, transferable between different contexts
+- Returned by activities, timers, and `Context.cancellable`
+
+#### `ResolvablePromise` (or `Trigger`)
+
+- Extends `ContextualPromise`
+- Provides additional public methods to resolve and reject itself
+- Useful for e.g waiting for a signal from WF main
+
+  ```ts
+  const unblocked = new ResolvablePromise<void>();
+  const signals = {
+    unblock(): void {
+      unblocked.resolve();
+    },
+  };
+
+  async function main() {
+    await unblocked;
+  }
+  ```
+
+### Open Questions
+
+#### Sharing promises between different contexts
+
+1. What happens when a timer or activity is created in a cancellable context and awaited in a non-cancellable context?
+
+```ts
+const p = sleep(3);
+await Context.nonCancellable(async () => {
+  // p is cancellable in a non cancellable context
+  await p;
+});
+```
+
+If the Workflow is cancelled, the user might be surprised if `await p` throws.
+
+2. What happens when a timer or activity was created in a non-cancellable context and awaited in a cancellable context?
+
+```ts
+let p: Promise<void> | undefined = undefined;
+await Context.nonCancellable(async () => {
+  p = sleep(3);
+});
+// p is non cancellable in a cancellable context
+await p;
+```
+
+If the Workflow is cancelled, the user might be surprised if `await p` does not throw.
+
+One approach would be to say that once a `ContextualPromise` is awaited in a non-cancellable context it becomes non-cancellable in all contexts. Another approach is the exact opposite where the promise becomes cancellable once awaited in a cancellable context.
+
+We should be able to support different cancellation policies for a single promise.
+With this approach, when a promise is cancelled while it is awaited in a both cancellable and non-cancellable contexts, an exception is thrown immediately in all cancellable contexts (assuming TRY_CANCEL cancellation type) but the activity does not get cancelled.
+
+The first 2 approaches are still surprising although they're predictable if they are well documented.
+The latter approach seems preferable to me and the least surprising. Not sure how feasable this is and if there are limitations with different cancellation types.
```

Comment:
Another possible (but probably crappy) option would be to simply blow up if a promise is awaited in different contexts, just explicitly prohibit the problem all-together, but not very javascript-y and would be much better solved at the type level if we could... :man_shrugging:

Reply:
Says the rust programmer :)
I don't think we can implement something like a borrow checker with Typescript.
I'd like to keep the ability to await on a context in multiple contexts if possible but it's definitely an option to just drop support for that.

Reply:
Added the approach you've mentioned to the doc.